### PR TITLE
Solver - add a wrapper for ``scipy`` L-BFGS solver

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
This adds a BFGS to the solvers' toolbox of `skglm`.
It proceeds as follows:

- [x] add wrapper around ``scipy.optimize.minimize(method="L-BFGS-B")``
- [x] unitest against `scikit-learn` L2-Logistic regression

### CODE REVIEW

File: skglm/penalties/separable.py [hunk 1]
```diff
@@ -557,3 +557,34 @@ def is_penalized(self, n_features):
...
+
+
+class L2(BasePenalty):
+    r""":math:`ell_2` penalty.
+
+    The penalty reads::
+
+    .. math::
+
+        \alpha xx \lVert w \rVert_2
```
Comment (mathurinm) [inline skglm/penalties/separable.py:569]: missing square?

Comment (mathurinm) [reply skglm/penalties/separable.py:569]: and division by 2

File: skglm/solvers/bfgs.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+import scipy.optimize
+from numpy.linalg import norm
+
+from skglm.solvers import BaseSolver
+from skglm.datafits import BaseDatafit
+
+
+class BFGS(BaseSolver):
```
Comment (mathurinm) [inline skglm/solvers/bfgs.py:9]: LBFGS should perform better a priori, if you can change

Comment (Badr-MOUFAD) [reply skglm/solvers/bfgs.py:9]: I wanted to conform with [`scipy` naming](https://docs.scipy.org/doc/scipy/reference/optimize.minimize-bfgs.html#optimize-minimize-bfgs) as well as the [reference, Chapter 6](https://link.springer.com/book/10.1007/978-0-387-40065-5).
Shall we rename it anyway?

Comment (mathurinm) [reply skglm/solvers/bfgs.py:9]: Need to check more carefully but I reckon scipy's BFGS uses BFGS, while I'm suggesting the limited memory variant, L-BFGS, that should be implemented here (in addition with box constraints that we ignore): https://docs.scipy.org/doc/scipy-0.18.1/reference/optimize.minimize-lbfgsb.html

**Comment by mathurinm** [comment]
Can you run it quickly on moderate size datasets in https://github.com/benchopt/benchmark_logreg_l2 ?

File: doc/api.rst [hunk 3]
```diff
@@ -36,6 +36,7 @@ Penalties
    IndicatorBox
    L0_5
    L1
+   L2
```
Comment (mathurinm) [inline doc/api.rst:39]: maintain lexigraphic order by putting it one line below

File: skglm/solvers/lbfgs.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+class LBFGS(BaseSolver):
+    """A wrapper for scipy L-BFGS solver.
+
+    Refer to `scipy L-BFGS-B <https://docs.scipy.org/doc/scipy/reference/optimize.
+    minimize-lbfgsb.html#optimize-minimize-lbfgsb>`_ documentation for details.
+
+    Parameters
+    ----------
+    max_iter : int, default 20
```
[the remark this answers is not in the document]
Comment (Badr-MOUFAD) [reply skglm/solvers/lbfgs.py:19]: 50?

**Comment by Badr-MOUFAD** [comment]
> Can you run it quickly on moderate size datasets in https://github.com/benchopt/benchmark_logreg_l2 ?

Here are [the benchmark figures](https://badr-moufad.github.io/share-benchmarks/bench-lbfgs/benchmark_logreg_l2_benchopt_run_2023-06-12_14h27m46.html) of skglm lbfgs vs. scikit-learn lbfgs

**Comment by mathurinm** [comment]
> > Can you run it quickly on moderate size datasets in https://github.com/benchopt/benchmark_logreg_l2 ?
> 
> Here are [the benchmark figures](https://badr-moufad.github.io/share-benchmarks/bench-lbfgs/benchmark_logreg_l2_benchopt_run_2023-06-12_14h27m46.html) of skglm lbfgs vs. scikit-learn lbfgs

I can't see the iterations on it, how many iterations does it take to converge?
On some datasets there seem to be a very small difference in final objective value (1e-6), it's not blocking but do you know the cause?

**Pull request merged**